### PR TITLE
TileLayer Roadmap

### PR DESCRIPTION
**Pull request by ibgreen**
This is a starting point for a roadmap for `TileLayer`, to get agreement / share awareness on technical directions.

I suspect we may be doing significant amount of engineering work on this layer (or a fork of it) during the 8.x release cycle, and there are several big PRs open on it, so a little bit of direction will surely not hurt.

For #

#### Change List
-

### CODE REVIEW

**Comment by chrisgervang** [comment]
Exciting proposal! The parent visibility tile cache issues become very pronounced when using the TerrainLayer since the meshes are intersecting and covering each other. 
Does anyone have implementation ideas on how to fix this? I've explored a bit in the tile cache, but haven't isolated the code causing the issue.

**Comment by kylebarron** [comment]
In the short/medium/long term, do we expect the `tileset2d` and `tileset3d` to stay as separate classes? Especially when looking at improvements such as high pitch/level of detail (#4482), those improvements would be helpful for both the 2D and 3D tiles cases. Or even as separate classes, there could be some areas where they share code. I suppose if `tileset2d` were moved to loaders.gl that could allow some code sharing.

**Comment by ibgreen** [comment]
> In the short/medium/long term, do we expect the tileset2d and tileset3d to stay as separate classes? Especially when looking at improvements such as high pitch/level of detail (#4482), those improvements would be helpful for both the 2D and 3D tiles cases. Or even as separate classes, there could be some areas where they share code.

Indeed, long-term, we have talked about having a single `TileLayer` in deck.gl that takes a `tileset` prop. Different subclasses of `Tileset` would implement loading schemes for 2D and 3D tiles.

We'd have to unify culling / tile-selection across 2D and 3D. If done right, that would generalize and simplify deck.gl, so it is a good direction, but needs to done carefully.

>  I suppose if tileset2d were moved to loaders.gl that could allow some code sharing.

Agree that this is the next logical step. Why has it not already happened? My major hesitations about moving 2d tileset into loaders.gl have been:
- impact on bundle size (if we put it in the tiles category, do we start pulling in a ton of 3d tiles code and helper math into the 2D layer)
- and also making sure the (relative) simplicity of the 2D tile system is not lost in unification - .

File: dev-docs/roadmaps/tile-layer-roadmap.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+## Proposals
+
+### Proposal: Breakout out `TileLayer` tile handling into `Tileset2D` class.
+
+- Would follow conventions/established pattern in `Tileset3D` in loaders.gl, making it easier for a maintainer to work on both layers.
+- Move `Tileset2D` to loaders.gl (easier maintenance)?
+- Use same API across `Tileset2D` and `Tileset3D`, e.g. for setting viewStates
+- Users who need to configure tile loading would not need to subclass the entire layer, just the `Tileset2D` class.
+
+### Proposal: data prop set to `Tileset` class
```
Comment (Pessimistress) [inline dev-docs/roadmaps/tile-layer-roadmap.md:60]: I think we have settled on offering another prop to override `Tileset2D` with a custom class.

**Review by Pessimistress — APPROVED** [review]

File: dev-docs/roadmaps/tile-layer-roadmap.md (hunk 1 above)
Comment (ibgreen) [reply dev-docs/roadmaps/tile-layer-roadmap.md:60]: Yes. This document is dated Jan 2020, I think it is OK to land it as-is, as a historical doc. I may refresh it part of ongoing work and will update this then.

File: dev-docs/roadmaps/tile-layer-roadmap.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+- No visual optimizations (e.g. load tiles from center out, blend in child-tiles, ...)
+
+
+### PRs
+
+At the time of writing, the following PRs were open. This roadmap tries to understand what these PRs want to achieve, then proposes a number of improvements to the `TileLayer`, then contains recommendations for these PRs in light of the proposals:
+
+- [`MVTTileLayer`](https://github.com/uber/deck.gl/pull/3935) - Highlights: Mapbox vector tile support - Worker loading of mapbox vector tiles - Better/Customized Tile Cache - Composite Geometry...
+- [Non-geospatial `TileLayer`](https://github.com/uber/deck.gl/pull/4117) - Provide the necessary hooks in to control coordinate mapping to tiles  `tileSize`, `tile2boundingBox`, `getTileIndices`,
+- [TerrainLayer](https://github.com/uber/deck.gl/pull/3984) - An example of `TileLayer` used to display terrain. Something we want to make easy (at least easier) for users.
```
Comment (kylebarron) [inline dev-docs/roadmaps/tile-layer-roadmap.md:49]:
```suggestion
- [`MVTTileLayer`](https://github.com/visgl/deck.gl/pull/3935) - Highlights: Mapbox vector tile support - Worker loading of mapbox vector tiles - Better/Customized Tile Cache - Composite Geometry...
- [Non-geospatial `TileLayer`](https://github.com/visgl/deck.gl/pull/4117) - Provide the necessary hooks in to control coordinate mapping to tiles  `tileSize`, `tile2boundingBox`, `getTileIndices`,
- [TerrainLayer](https://github.com/visgl/deck.gl/pull/3984) - An example of `TileLayer` used to display terrain. Something we want to make easy (at least easier) for users.
```

File: dev-docs/roadmaps/tile-layer-roadmap.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+### Proposal: Implement proper tile replacement
+
+- Simple mode: Show parent tiles until all child tiles loaded (must be done whenever opacity != 1)
+- Advanced mode: Blend child tiles into parent tiles?
+
+## Way forward for Open PRs
+
+To help providing feedback in the PRs themselves, it may help discuss how they fit into the TileLayer roadmap.
+
+### PR [`MVTTileLayer`](https://github.com/uber/deck.gl/pull/3935)
```
Comment (kylebarron) [inline dev-docs/roadmaps/tile-layer-roadmap.md:98]:
```suggestion
### PR [`MVTTileLayer`](https://github.com/visgl/deck.gl/pull/3935)
```

File: dev-docs/roadmaps/tile-layer-roadmap.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+Proposed feedback:
+- Worker Thread loading of [MVT](https://github.com/mapbox/vector-tile-js) (mapbox vector tiles) - Remove, and build/use new `@loaders.gl/mvt` (or `@loaders.gl/mapbox-vector-tile`) module.
+- Tile Cache - Generalize and put in core TileLayer (or possibly share TileCache with `Tileset3D` in loaders.gl)
+- Composite Geometry - do not merge, keep this on application side, just make sure `TileLayer` offers necessary hooks that app needs.
+
+### PR [Non-geospatial `TileLayer`](https://github.com/uber/deck.gl/pull/4117)
```
Comment (kylebarron) [inline dev-docs/roadmaps/tile-layer-roadmap.md:105]:
```suggestion
### PR [Non-geospatial `TileLayer`](https://github.com/visgl/deck.gl/pull/4117)
```

File: dev-docs/roadmaps/tile-layer-roadmap.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+Proposed feedback:
+- Do not create a separate base class
+- Provide the necessary hooks in the standard `TileLayer` (or `Tileset2D` class) to control coordinate mapping to tiles
+   - `tileSize`,
+   - `tile2boundingBox`,
+   - `getTileIndices`,
+- Add test data! TileLayer is likely going to be heavily refactored over the next two years and deck.gl is unlikely to be able to keep non-geospatial support working without tests and examples.
+
+### PR [TerrainLayer WIP](https://github.com/uber/deck.gl/pull/3984)
```
Comment (kylebarron) [inline dev-docs/roadmaps/tile-layer-roadmap.md:115]:
```suggestion
### PR [TerrainLayer WIP](https://github.com/visgl/deck.gl/pull/3984)
```

**Review by kylebarron — APPROVED** [review]
Links should be updated to the `visgl` repository.

**Pull request merged**